Adding `cmake` as an input causes the stdenv to try to build the entire expression with cmake instead of cargo

Having `cmake` in Nix dependencies breaks the build with the following error message:-

```
CMake Error: The source directory "/build/dummy-src" does not appear to contain CMakeLists.txt.
```

I tried `depsBuildBuild`, `nativeBuildInputs`, `buildInputs` and even `depsTargetTarget`. Putting `cmake` in any of those for a project that doesn't contain `CMakeLists.txt` causes it to abort with that message. I'm not sure if this is the expected behavior or not but it's surprising.

I was able to get around this by using the environment variable `CMAKE = "${pkgs.cmake}/bin/cmake"`. ~~However, doing this seems to override `gcc` from `stdenv` and uses the latest `gcc` instead. I suspect that's why `overrideScope'` didn't work for me in this discussion https://github.com/ipetkov/crane/discussions/99. I even ran into this as well when I tried the PR for ergonomic `stdenv` here https://github.com/ipetkov/crane/discussions/99#discussioncomment-3655905. I had to initialize a new project with no Nix dependencies to verify that the new PR was in fact working.~~

Edit: I'm not so sure about this anymore. Commenting out the environment variable still brings in the newer `gcc`. This is strange because it was working fine with `naersk` before switching to `crane`.

Interesting. I tried in my project, by just:


```
> git diff
diff --git a/flake.nix b/flake.nix
index b2b73c736..0af864fdc 100644
--- a/flake.nix
+++ b/flake.nix
@@ -141,6 +141,7 @@
           src = filterWorkspaceFiles ./.;
 
           buildInputs = with pkgs; [
+            cmake
             clang
             gcc
             openssl

```

and indeed I've got:


```
> nix build .#workspaceBuild
warning: Git tree '/home/dpc/lab/fedimint' is dirty
error: builder for '/nix/store/4fvp4aqkjx882blhbrllvmljqzv3rz1f-workspace-deps-deps-0.0.1.drv' failed with exit code 1;
       last 10 log lines:
       > unpacking source archive /nix/store/31plhj7fld43www31lhf7dc6bx1xn3my-dummy-src
       > source root is dummy-src
       > patching sources
       > Executing configureCargoCommonVars
       > configuring
       > will append /build/dummy-src/.cargo-home/config.toml with contents of /nix/store/a4p08jfqw8p1czhxjip0qldhyn14y45b-vendor-cargo-deps/config.toml
       > fixing cmake files...
       > cmake flags: -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_SKIP_BUILD_RPATH=ON -DBUILD_TESTING=OFF -DCMAKE_INSTALL_LOCALEDIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/share/locale -DCMAKE_INSTALL_LIBEXECDIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/libexec -DCMAKE_INSTALL_LIBDIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/lib -DCMAKE_INSTALL_DOCDIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/share/doc/workspace-deps-deps -DCMAKE_INSTALL_INFODIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/share/info -DCMAKE_INSTALL_MANDIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/share/man -DCMAKE_INSTALL_OLDINCLUDEDIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/include -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/include -DCMAKE_INSTALL_SBINDIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/sbin -DCMAKE_INSTALL_BINDIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/bin -DCMAKE_INSTALL_NAME_DIR=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1/lib -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_OSX_SYSROOT= -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_STRIP=/nix/store/ag2bpk0lzjvj409znklrz5krkpc5imzs-gcc-wrapper-11.3.0/bin/strip -DCMAKE_RANLIB=/nix/store/rq6bh3qfrqnyqwik0w3q6w180zg3w2pa-binutils-2.38/bin/ranlib -DCMAKE_AR=/nix/store/rq6bh3qfrqnyqwik0w3q6w180zg3w2pa-binutils-2.38/bin/ar -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/qnjbz35ra1rzwph2lizl49n5gy3rgvhb-workspace-deps-deps-0.0.1
       > CMake Error: The source directory "/build/dummy-src" does not appear to contain CMakeLists.txt.
       > Specify --help for usage, or press the help button on the CMake GUI.
       For full logs, run 'nix log /nix/store/4fvp4aqkjx882blhbrllvmljqzv3rz1f-workspace-deps-deps-0.0.1.drv'.
error: 1 dependencies of derivation '/nix/store/b88mh6xijxj1i3mp3y52bjhar3hly449-workspace-build-0.0.1.drv' failed to build
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ## Unreleased
 
 ## Changed
+* **Breaking**: `mkCargoDerivation` now includes a default `configurePhase`
+  which does nothing but run the `preConfigure` and `postConfigure` hooks. This
+  is done to avoid breaking builds by including puts happen to have setup-hooks
+  which try to claim the configure phase (such as `cmake`). To get the old
+  behavior back, set `configurePhase = null;` in the derivation.
 * `mkCargoDerivation` (along with any of its callers like `cargoBuild`,
   `buildPackage`, etc.) now accept a `stdenv` argument which will override the
   default environment (coming from `pkgs.stdenv`) for that particular derivation

--- a/checks/default.nix
+++ b/checks/default.nix
@@ -146,6 +146,13 @@ myPkgs // {
     src = ./simple;
     stdenv = pkgs.gcc12Stdenv;
   };
+  # https://github.com/ipetkov/crane/issues/104
+  simpleWithCmake = myLib.buildPackage {
+    src = ./simple;
+    nativeBuildInputs = with pkgs; [
+      cmake
+    ];
+  };
 
   simple-nonflake = (import ../default.nix {
     inherit pkgs;

--- a/docs/API.md
+++ b/docs/API.md
@@ -631,6 +631,9 @@ This is a fairly low-level abstraction, so consider using `buildPackage` or
 * `checkPhase`: the commands used by the check phase of the derivation
   - Default value: the check phase will run `preCheck` hooks, log and evaluate
     `checkPhaseCargoCommand`, and run `postCheck` hooks
+* `configurePhase`: the commands used by the configure phase of the derivation
+  - Default value: the configure phase will run `preConfigureHooks` hooks, then
+    run `postConfigure` hooks
 * `doInstallCargoArtifacts`: controls whether cargo's `target` directory should
   be copied as an output
   - Default value: `true`

--- a/lib/buildPackage.nix
+++ b/lib/buildPackage.nix
@@ -52,6 +52,7 @@ in
 
   buildPhase = args.buildPhase or ''
     runHook preBuild
+    cargo --version
     ${buildPhaseCargoCommand}
     runHook postBuild
   '';

--- a/lib/mkCargoDerivation.nix
+++ b/lib/mkCargoDerivation.nix
@@ -64,6 +64,12 @@ chosenStdenv.mkDerivation (cleanedArgs // {
     runHook postCheck
   '';
 
+  configurePhase = args.configurePhase or ''
+    runHook preConfigure
+    echo default configurePhase, nothing to do
+    runHook postConfigure
+  '';
+
   installPhase = args.installPhase or ''
     runHook preInstall
     ${installPhaseCommand}

